src/adapter/src/optimize/subscribe.rs

```diff
@@ -0,0 +1,322 @@
+/// optimization result with a resolved timestamp.
+///
+/// The actual timestamp value is set in the [`MirDataflowDescription`] of the
+/// surrounding [`GlobalMirPlan`] when we call `resolve()`.
+#[derive(Clone)]
+pub struct Resolved;
+
+/// The (final) result after MIR ⇒ LIR lowering and optimizing the resulting
+/// `DataflowDescription` with `LIR` plans.
+pub struct GlobalLirPlan {
+    df_desc: LirDataflowDescription,
+    df_meta: DataflowMetainfo,
+}
+
+impl OptimizeSubscribe {
+    pub fn new(
+        catalog: Arc<Catalog>,
+        compute_instance: ComputeInstanceSnapshot,
+        transient_id: GlobalId,
+        conn_id: ConnectionId,
+        with_snapshot: bool,
+        up_to: Option<Timestamp>,
+        config: OptimizerConfig,
+    ) -> Self {
+        Self {
+            typecheck_ctx: empty_context(),
+            catalog,
+            compute_instance,
+            transient_id,
+            conn_id,
+            with_snapshot,
+            up_to,
+            config,
+        }
+    }
+}
+
+impl Optimize<SubscribeFrom> for OptimizeSubscribe {
+    type To = GlobalMirPlan<Unresolved>;
+
+    fn optimize(&mut self, plan: SubscribeFrom) -> Result<Self::To, OptimizerError> {
+        let sink_name = format!("subscribe-{}", self.transient_id);
+
+        let (df_desc, df_meta) = match plan {
+            SubscribeFrom::Id(from_id) => {
+                let from = self.catalog.get_entry(&from_id);
+                let from_desc = from
+                    .desc(
+                        &self
+                            .catalog
+                            .state()
+                            .resolve_full_name(from.name(), Some(&self.conn_id)),
+                    )
+                    .expect("subscribes can only be run on items with descs")
+                    .into_owned();
+
+                // Make SinkDesc
+                let sink_id = self.transient_id;
+                let sink_desc = ComputeSinkDesc {
+                    from: from_id,
+                    from_desc,
+                    connection: ComputeSinkConnection::Subscribe(SubscribeSinkConnection::default()),
+                    with_snapshot: self.with_snapshot,
+                    up_to: self.up_to.map(Antichain::from_elem).unwrap_or_default(),
+                    // No `FORCE NOT NULL` for subscribes
+                    non_null_assertions: vec![],
+                };
+
+                let mut df_builder =
+                    DataflowBuilder::new(self.catalog.state(), self.compute_instance.clone());
+
+                let (df_desc, df_meta) =
+                    df_builder.build_sink_dataflow(sink_name, sink_id, sink_desc)?;
+
+                (df_desc, df_meta)
+            }
+            SubscribeFrom::Query { expr, desc } => {
+                // TODO: expr: HirRelationExpr
+                // HIR ⇒ MIR lowering and decorrelation
+                // let expr = expr.lower()?;
```

Comment:
I don't understand this comment. Assuming it's not stray code, could you add more words to explain what this TODO is about?

Reply:
Yes. This is because the `Query` variant comes with a decorrelated `expr`. In the long term we want to change all `Plan` variants to host `HirRelationExpr` instead of `MirRelationExpr` and move decorrelation to the optimizer.